tidb-binlog/pump_client/client.go

```diff
@@ -0,0 +1,410 @@
+	}
+
+	c.Pumps.Unlock()
+}
+
+// updatePump update pump's status, and return whether pump's IsAvaliable should be changed.
+func (c *PumpsClient) updatePump(status *node.Status) (pump *PumpStatus, avaliableChanged, avaliable bool) {
+	var ok bool
+	c.Pumps.Lock()
+	if pump, ok = c.Pumps.Pumps[status.NodeID]; ok {
+		if pump.Status.State != status.State {
+			if status.State == node.Online {
+				avaliableChanged = true
+				avaliable = true
+			} else if pump.Status.State == node.Online {
+				avaliableChanged = true
+				avaliable = false
+			}
+		}
+		pump.Status = *status
+	}
+	c.Pumps.Unlock()
+
+	return
+}
+
+// removePump removes a pump.
+func (c *PumpsClient) removePump(nodeID string) {
+	c.Pumps.Lock()
+	if pump, ok := c.Pumps.Pumps[nodeID]; ok {
+		pump.closeGrpcClient()
+	}
+	delete(c.Pumps.Pumps, nodeID)
+	delete(c.Pumps.UnAvaliablePumps, nodeID)
+	delete(c.Pumps.AvaliablePumps, nodeID)
+	c.Selector.SetPumps(copyPumps(c.Pumps.AvaliablePumps))
+	c.Pumps.Unlock()
+}
+
+// exist returns true if pumps client has pump matched this nodeID.
+func (c *PumpsClient) exist(nodeID string) bool {
+	c.Pumps.RLock()
+	_, ok := c.Pumps.Pumps[nodeID]
+	c.Pumps.RUnlock()
+	return ok
+}
+
+// watchStatus watchs pump's status in etcd.
+func (c *PumpsClient) watchStatus() {
+	defer c.wg.Done()
+	rootPath := path.Join(node.DefaultRootPath, node.NodePrefix[node.PumpNode])
+	rch := c.EtcdRegistry.WatchNode(c.ctx, rootPath)
+	for {
+		select {
+		case <-c.ctx.Done():
+			log.Info("[pumps client] watch status finished")
+			return
+		case wresp := <-rch:
+			for _, ev := range wresp.Events {
+				log.Debugf("[pumps client] watch etcd event type:%s, key: %q, value: %q", ev.Type, ev.Kv.Key, ev.Kv.Value)
+				status := &node.Status{}
+				err := json.Unmarshal(ev.Kv.Value, &status)
+				if err != nil {
+					log.Errorf("[pumps client] unmarshal pump status %q failed", ev.Kv.Value)
+					continue
+				}
+
+				switch ev.Type {
+				case mvccpb.PUT:
+					if !c.exist(status.NodeID) {
+						c.addPump(NewPumpStatus(status), true)
+						continue
+					}
+
+					pump, avaliableChanged, avaliable := c.updatePump(status)
+					if avaliableChanged {
+						c.setPumpAvaliable(pump, avaliable)
+					}
+
+				case mvccpb.DELETE:
+					// now will not delete pump node in fact, just for compatibility.
+					nodeID := node.AnalyzeNodeID(string(ev.Kv.Key))
+					c.removePump(nodeID)
+				}
+			}
+		}
+	}
+}
+
+// detect send detect binlog to pumps with online state in UnAvaliablePumps,
+func (c *PumpsClient) detect() {
+	defer c.wg.Done()
+	for {
+		select {
+		case <-c.ctx.Done():
+			log.Infof("[pumps client] heartbeat finished")
+			return
+		default:
+			// send detect binlog to pump, if this pump can return response without error
+			// means this pump is avaliable.
+			needCheckPumps := make([]*PumpStatus, 0, len(c.Pumps.UnAvaliablePumps))
+			checkPassPumps := make([]*PumpStatus, 0, 1)
+			req := &pb.WriteBinlogReq{ClusterID: c.ClusterID, Payload: nil}
```

Comment:
so, pump need to handle nil payload specially just return success?

Reply:
yes